### DISCUSSION TITLE
why is CMAKE_CROSSCOMPILING set to TRUE?

### DISCUSSION THREAD

**Discussion by bowie7070** [MarkSchofield/Toolchain#105] [Q&A]
It seems to be the case that we are not cross-compiling and having CMAKE_CROSSCOMPILING setting it to OFF resolves some problems I had (I recollect it was compiling python in vcpkg but I'll need to double check if anyone wants to know accurately). What is the reason that CMAKE_CROSSCOMPILING is set to TRUE?

**Comment by MarkSchofield** [comment]
The documentation for [cmake-toolchains](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) suggests that if you're specifying a toolchain file, then you're cross-compiling:

> CMake uses a toolchain of utilities to compile, link libraries and create archives, and other tasks to drive the build. The toolchain utilities available are determined by the languages enabled. In normal builds, CMake automatically determines the toolchain for host builds based on system introspection and defaults. In cross-compiling scenarios, a toolchain file may be specified with information about compiler and utility paths.

i.e. 'normal builds' don't use a toolchain. If you're using a toolchain file, then you are cross-compiling.

**Reply by bowie7070** [reply]
I see why you might think that from what you have read but that's not the definition of cross compiling. Cross compiling refers to compiling to produce an executable or library for a system (the target) that is different than the one that is doing the compilation (the host). You can use a toolchain for y without doing cross compiling and you can do cross compiling without using a toolchain file. 



https://cmake.org/cmake/help/book/mastering-cmake/chapter/Cross%20Compiling%20With%20CMake.html#:~:text=Cross%2Dcompiling%20a%20piece%20of,run%20on%20a%20different%20system.

**Reply by MarkSchofield** [reply]
Ah, gotcha. TBH, [this snippet](https://github.com/Kitware/CMake/blob/b346afce4fc977d875c613334ebb5365cd647453/Modules/CMakeDetermineSystem.cmake#L156C1-L182C8) seems to be logic that CMake uses to define cross-compiling. So it seems like the toolchains shouldn't be setting `CMAKE_CROSSCOMPILING`, and they shouldn't be setting `CMAKE_SYSTEM_NAME`, so that CMake will default it to `CMAKE_HOST_SYSTEM_NAME`, and CMake won't define `CROSS_COMPILING`.

**Reply by MarkSchofield** [reply]
Hmmm... But the same page that you linked, goes on to say that '[the `CMAKE_SYSTEM_NAME`] value is mandatory'. I think it means that 'the `CMAKE_SYSTEM_NAME` value is mandatory _for toolchain files to set_'. In which case, setting `CMAKE_SYSTEM_NAME` causes CMake to implicitly set `CMAKE_CROSSCOMPILING` to `true`. And that agrees with my interpretation of [the cmake-toolschains page](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) - by specifying a toolchain, then `CMAKE_CROSSCOMPILING` should be true.

TBH, your definition makes sense, and the code I linked ([here](https://github.com/Kitware/CMake/blob/b346afce4fc977d875c613334ebb5365cd647453/Modules/CMakeDetermineSystem.cmake#L156C1-L182C8)) makes sense from that definition. I pulled together a branch with the straight-forward changes - [mschofie/cross-compiling](https://github.com/MarkSchofield/WindowsToolchain/compare/main...mschofie/cross-compiling) - but it's still not clear to me what the Right Thing to do is.

**Reply by bowie7070** [reply]
> I think it means that 'the CMAKE_SYSTEM_NAME value is mandatory for toolchain files to set'

I can see why you might reach that conclusion but it rather it is that CMAKE_SYSTEM_NAME value is mandatory when cross-compiling because cmake can't guess what platform you are targetting. It is certainly not the case that CMAKE_SYSTEM_NAME must be set in general because "CMAKE_SYSTEM_NAME is by default set to the same value as the [CMAKE_HOST_SYSTEM_NAME](https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_NAME.html#variable:CMAKE_HOST_SYSTEM_NAME) variable so that the build targets the host system." The appropriate value for CMAKE_HOST_SYSTEM_NAME is able to be detected automatically by cmake.

https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html

The use of a toolchain file does not imply that you are cross compiling. A toolchain file is typically used any time the default toolchain that cmake would find is not suitable, such as when you want to use a specific version of a compiler.

When cross-compiling you almost certainly want a toolchain file because cmake cannot guess the many details that are known to the developer. However, toolchain files are frequently used when not cross-compiling. For example, it is necessary to use a toolchain file when using vcpkg in manifest mode.

**Reply by bowie7070** [reply]
> I pulled together a branch with the straight-forward changes - [mschofie/cross-compiling](https://github.com/MarkSchofield/WindowsToolchain/compare/main...mschofie/cross-compiling) - but it's still not clear to me what the Right Thing to do is.

The changes in this PR look good. But I've only looked at them quickly and I haven't tested them.

**Comment by MarkSchofield** [comment]
Fixed with #124.
